### DISCUSSION TITLE
Need any AI help?

### DISCUSSION THREAD

**Discussion by metaylor** [ihhub/fheroes2#3435] [Q&A]
Hi,

Thanks for everyone's work on this awesome project!

I'm computer science professor specializing in artificial intelligence. I haven't programmed in years, but if there's something I could do to help by talking through an approach, answering AI/planning/machine learning questions, suggesting algorithms, etc., let me know?

Best,
Matt

**Comment by ihhub** [comment]
Hi @metaylor , thank you very much for the interest and offering the help of yours!

First of all, our main issue in AI is the logic related to kingdom development on the map such as hero's movements, castle building and planning actions in advance. Right now we have a weight-based model where the hero's actions are based on the priority of the objects. Priority is calculated based on multiple parameters and this is just a raw number. Of course it doesn't work properly in most of the cases but this is just a foundation.

Secondly. the most of the code is written by @idshibanov and I'm sure he has many plans for enhancement of the AI. He could share his ideas here or you could join our Discord server.

Additionally, we're bound by some bugs in the game's logic so we have to fix some of them as well.

In the future we're planning to have AI prediction logic based on the calculation of every possible outcome of action (applicable only for the battle as of now). Of course the more possibilities the more complex calculations become so we have to find a golden approach to balance between quality of the algorithm and speed.

Lastly, we have a [separate topic](https://github.com/ihhub/fheroes2/issues/2216) where some of machine learning questions were raised and answered :)

**Reply by metaylor** [reply]
Nice! Are the weights learned or hand coded? I bet you could optimize them per difficulty level, per map, etc. It seems like it could be quite hard to balance short term and long term goals (e.g., buying a few more units for a battle 1-2 turns away vs. constructing a new building that won't have new units until next week). 

Could you or @idshibanov please point me to the discord? I can bug you more there.

Trading off speed and optimality is the whole challenge of AI - I love it! Ideally you'd be able to build an AI that has super-human performance (without cheating), and then figure out how to scale it back to more appropriate difficulty levels. Or, as a first step, you can always cheat (different starting resources, full observability, weighted dice, better odds against NPCs, etc.)

**Reply by metaylor** [reply]
Also, it looks like I can donate via PayPal or Pateron. Do you have a preference (e.g., does one take a small cut of the amount)?

**Reply by idshibanov** [reply]
Hi @metaylor,

Current AI code was definitely built on "playable game first, improvements later" principle. There's a ton to improve, including weights which are hardcoded at the moment. I based task priorities on movement costs, i.e. how far a hero is ready to move to ignore what's in front of him.

Optimizing weights based on map size and tiles explored is a great idea. I believe a straightforward way to add ML to the game is to let it define the object weights. What's challenging to me is how we can build an adequate training set for the model or how to asses action outcomes when it's not immediate (captured a mine, lost battle but weakened enemy, etc).

Biggest issue in strategic planning right now is that every AI hero essentially acts on its own, looking for things to do. There are few optimizations in place to force them co-operate (objects "locking", army exchange), but we have to go a level above. I want a Kingdom manager implemented which will define global/regional tasks and assign them to heroes based on hero role/location. We also have to teach AI to set up logistic chain in order to reinforce main hero on the front lines.

Our development discord link is https://discord.gg/xF85vbZ. It's often faster to get a response there.

Cheers!

**Reply by ihhub** [reply]
> Also, it looks like I can donate via PayPal or Pateron. Do you have a preference (e.g., does one take a small cut of the amount)?

Hi @metaylor , no, we don't have any preference. It's totally up to you to decide. PayPal is more convenient for one time donation while Patreon is usually used for monthly deductions. The choice is totally yours🙂
